src/password.ml

```diff
@@ -1,41 +1,41 @@
-open Digestif
+open Cryptokit
 
-let random_salt size =
-  let mask = 255L in
-  let salt = Bytes.create size in
-  for i = 0 to size / 8 do
-    let random_int = Random.int64 Int64.max_int in
-    for j = 0 to min (size - (8 * i)) 8 - 1 do
-      Bytes.set salt
-        ((8 * i) + j)
-        (Char.chr
-           Int64.(to_int (logand mask (shift_right_logical random_int j))))
-    done
-  done;
-  Bytes.to_string salt
+let random_salt size = Random.(string secure_rng size)
 
 (* Return Result.Ok (update, auth) where auth is true if the auth is successful
    and update is true if the hash need to be updated *)
 let check stored_hash password =
+  let hash_decode = Hexa.decode () in
   match String.split_on_char '$' stored_hash with
   | [ ""; "1"; password_hash ] -> (
-      match MD5.of_hex_opt password_hash with
-      | Some stored_digest ->
-          let digest = MD5.digest_string password in
-          Result.Ok (true, MD5.equal stored_digest digest)
-      | None -> Result.Error "Failure reading stored hash" )
+      try
+        let hash_function = Hash.md5 () in
+        hash_decode#put_string password_hash;
+        hash_decode#finish;
+        hash_function#add_string password;
+        Result.Ok (true, string_equal hash_decode#get_string hash_function#result)
+      with Error _ -> Result.Error "Failure reading stored hash")
   | [ ""; "8"; salt; password_hash ] -> (
-      match SHA3_512.of_hex_opt password_hash with
-      | Some stored_digest ->
-          let digest =
-            SHA3_512.digest_string (password ^ Hex.to_string (`Hex salt))
-          in
-          Result.Ok (false, SHA3_512.equal stored_digest digest)
-      | None -> Result.Error "Failure reading stored hash" )
+      try
+        let salt_decode = Hexa.decode () in
+        salt_decode#put_string salt;
+        salt_decode#finish;
+        let hash_function = Hash.sha3 512 in
+        hash_decode#put_string password_hash;
+        hash_decode#finish;
+        hash_function#add_string (password^salt_decode#get_string);
+        Result.Ok (true, string_equal hash_decode#get_string hash_function#result)
```

Comment:
Shouldn't `true` be `false` here?

Reply:
Yes, thank you for catching this mistake.